### PR TITLE
refactor Token Wrapper out

### PR DESCRIPTION
**Pull request by hav-noms**
- [x] Refactor out the TokenWrapper
- [x] Implement IStakingRewards

### CODE REVIEW

File: contracts/StakingRewards.sol [hunk 1]
```diff
@@ -98,25 +71,29 @@ contract StakingRewards is TokenWrapper, RewardsDistributionRecipient {
...
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    function stake(uint256 amount) public nonReentrant updateReward(msg.sender) {
         require(amount > 0, "Cannot stake 0");
-        super.stake(amount);
+        _totalSupply = _totalSupply.add(amount);
+        _balances[msg.sender] = _balances[msg.sender].add(amount);
+        stakingToken.safeTransferFrom(msg.sender, address(this), amount);
```
Comment (justinjmoses) [inline contracts/StakingRewards.sol:84]: qu: is it worth trying to do `transferFromAndSettle` if possible?

Comment (justinjmoses) [reply contracts/StakingRewards.sol:84]: Nvm - having synths directly in here isn't the primary use-case.

File: contracts/StakingRewards.sol [hunk 2]
```diff
@@ -5,79 +5,52 @@ import "openzeppelin-solidity-2.3.0/contracts/math/SafeMath.sol";
...
-        rewardPerTokenStored = rewardPerToken();
-        lastUpdateTime = lastTimeRewardApplicable();
-        if (account != address(0)) {
-            rewards[account] = earned(account);
-            userRewardPerTokenPaid[account] = rewardPerTokenStored;
-        }
-        _;
+    /* ========== VIEWS ========== */
+
+    function totalSupply() public view returns (uint256) {
```
Comment (justinjmoses) [inline contracts/StakingRewards.sol:48]: Can you make this and the others all `external`?

File: contracts/interfaces/IStakingRewards.sol [hunk 3]
```diff
@@ -2,21 +2,22 @@ pragma solidity >=0.4.24;
 
 
 interface IStakingRewards {
+    // Views
```
Comment (justinjmoses) [inline contracts/interfaces/IStakingRewards.sol:5]: can you add `totalSupply()` and `balanceOf(...)` here as well? (If it was a full ERC20 token then I'd suggest not to as it should instead be cast to `IERC20` by third parties, but as it's partial I think it makes sense to include them, no?)

Comment (hav-noms) [reply contracts/interfaces/IStakingRewards.sol:5]: Agree

File: contracts/StakingRewards.sol (hunk 2 above)
Comment (hav-noms) [reply contracts/StakingRewards.sol:48]: totalsupply and balance of are used internally also

Comment (hav-noms) [reply contracts/StakingRewards.sol:48]: Can only do it on stake and withdraw

File: contracts/interfaces/IStakingRewards.sol (hunk 3 above)
Comment (hav-noms) [reply contracts/interfaces/IStakingRewards.sol:5]: RFAL

File: contracts/StakingRewards.sol [hunk 4]
```diff
@@ -98,25 +71,29 @@ contract StakingRewards is TokenWrapper, RewardsDistributionRecipient {
         return balanceOf(account).mul(rewardPerToken().sub(userRewardPerTokenPaid[account])).div(1e18).add(rewards[account]);
     }
 
-    // stake visibility is public as overriding LPTokenWrapper's stake() function
-    function stake(uint256 amount) public updateReward(msg.sender) {
+    function getRewardForDuration() public view returns (uint256) {
```
Comment (jacko125) [inline contracts/StakingRewards.sol:74]: the helper view here can be external.

Comment (hav-noms) [reply contracts/StakingRewards.sol]: done

File: contracts/StakingRewards.sol (hunk 2 above)
Comment (justinjmoses) [reply contracts/StakingRewards.sol:48]: but you can refactor those usages to  `_totalSupply` and `_balances` no?

File: contracts/interfaces/IStakingRewards.sol (hunk 3 above)
Comment (justinjmoses) [reply contracts/interfaces/IStakingRewards.sol:5]: thanks

File: contracts/StakingRewards.sol (hunk 2 above)
Comment (hav-noms) [reply contracts/StakingRewards.sol:48]: of course! done

**Comment by hav-noms** [comment]
RFAL

**Pull request merged**